### DISCUSSION TITLE
Support STM32U0

### DISCUSSION THREAD

**Discussion by sebfr26** [hathach/tinyusb#2667] [Ideas]
Hi together,

Is anyone planning to add support for the new STM32U0?

**Comment by hathach** [comment]
currently no plan, since I don't have the dev board just yet.

**Reply by sebfr26** [reply]
Thank you for your reply.

We are planning to use the STM32U0 with USB in future. We already have Evalboards. How can we support the implemenation? Since we are pretty new to this topic I'm not sure if it make sense to make the implementation by ourselves. On the otherside I hope the U0 is quite similar to the other STMs so that the effort is not that high.

**Reply by HiFiPhile** [reply]
Hi, they are already many MCU families implemented by contributors.

As U0 use the same DCD as G0, I suggest you to start with G0 and make modifications.
Basically a port needs several parts:
1. Add U0's ID in https://github.com/hathach/tinyusb/blob/ba2f2299c34491aa8c354e00fa3f9d222e30db89/src/tusb_option.h#L93
2. Add define to choose DCD code named "FSDEV_STM32" in https://github.com/hathach/tinyusb/blob/ba2f2299c34491aa8c354e00fa3f9d222e30db89/src/common/tusb_mcu.h#L237
3. Add U0 HAL driver for BSP usage in https://github.com/hathach/tinyusb/blob/master/tools/get_deps.py
4. Create BSP in hw/bsp/stm32u0, you can reference G0's structure https://github.com/hathach/tinyusb/tree/master/hw/bsp/stm32g0
5. Build and test examples.


_From your ID are you in France ?_

**Reply by sebfr26** [reply]
Thank you.
We will try to get this running in the next weeks.

I'm from Germany...

**Comment by JimmyPedersen** [comment]
@sebfr26 Have you tried getting it to run on the stm32u0 yet? Any luck?
Looking for this to as the 'new' usb stack is hard to get working well WITHOUT RTOS ( threadx)

**Reply by Maerdl** [reply]
I recently placed a pull request : https://github.com/hathach/tinyusb/pull/2824
But there are no examples.

We got that run at least as cdc device with success. U0 uses more or less the same usb core than U5.

I forked that in my repo: https://github.com/Maerdl/tinyusb/tree/master
